igvc_gazebo/nodes/ground_truth/main.cpp

```diff
@@ -25,28 +37,38 @@ void groundTruthCallback(const nav_msgs::Odometry::ConstPtr& msg)
   {
     g_og_pose.pose = msg->pose;
     g_og_pose.header = msg->header;
-    g_og_pose.pose.pose.position.x = msg->pose.pose.position.x;
-    g_og_pose.pose.pose.position.y = msg->pose.pose.position.y;
-    ROS_INFO_STREAM("setting g_og_pose to " << g_og_pose.pose.pose.position.x << ", "
-                                            << g_og_pose.pose.pose.position.y);
+    g_og_pose.pose.pose.position.x = msg->pose.pose.position.x + x_distribution(engine);
+    g_og_pose.pose.pose.position.y = msg->pose.pose.position.y + y_distribution(engine);
+    g_og_pose.pose.pose.position.z = msg->pose.pose.position.z + z_distribution(engine);
+    ROS_INFO_STREAM("setting g_og_pose to " << g_og_pose.pose.pose.position.x << ", " << g_og_pose.pose.pose.position.y
+                                            << ", " << g_og_pose.pose.pose.position.z);
   }
   else
   {
     nav_msgs::Odometry result;
     result.pose = msg->pose;
 
     // use the initial location as an offset (makes the starting location 0, 0)
-    result.pose.pose.position.x = msg->pose.pose.position.x - g_og_pose.pose.pose.position.x;
-    result.pose.pose.position.y = msg->pose.pose.position.y - g_og_pose.pose.pose.position.y;
-
+    result.pose.pose.position.x = msg->pose.pose.position.x - g_og_pose.pose.pose.position.x + x_distribution(engine);
+    result.pose.pose.position.y = msg->pose.pose.position.y - g_og_pose.pose.pose.position.y + y_distribution(engine);
+    result.pose.pose.position.z = msg->pose.pose.position.z - g_og_pose.pose.pose.position.z + z_distribution(engine);
     result.twist = msg->twist;
 
     // set up the correct header
     result.header = msg->header;
     result.child_frame_id = "base_footprint";
     result.header.frame_id = "odom";
 
-    tf::Quaternion quat;
+    tf::Quaternion quat(msg->pose.pose.orientation.x, msg->pose.pose.orientation.y, msg->pose.pose.orientation.z,
+                        msg->pose.pose.orientation.w);
+    tf::Matrix3x3 m(quat);
+    double roll, pitch, yaw;
+    m.getRPY(roll, pitch, yaw);
+    roll = roll + roll_distribution(engine);
+    pitch = pitch + pitch_distribution(engine);
+    yaw = yaw + yaw_distribution(engine);
```

Comment:
Currently, this doesn't change the published message. You want something like below which constructs a new Quaternion using the RPY values plus noise.
```suggestion
    result.pose.pose.orientation = tf::createQuaternionMsgFromRollPitchYaw(
        roll + roll_distribution(engine), pitch + pitch_distribution(engine), yaw + yaw_distribution(engine));
```